sycl/include/CL/sycl/atomic.hpp

```diff
@@ -274,18 +280,36 @@ class atomic {
 #endif
   }
 
-  T fetch_add(T Operand, memory_order Order = memory_order::relaxed) {
-    __SYCL_STATIC_ASSERT_NOT_FLOAT(T);
+  template <typename T2 = T>
+  detail::enable_if_t<std::is_integral<T2>::value, T>
+  fetch_add(T Operand, memory_order Order = memory_order::relaxed) {
     return __spirv_AtomicIAdd(
         Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
   }
 
-  T fetch_sub(T Operand, memory_order Order = memory_order::relaxed) {
+  template <typename T2 = T>
+  detail::enable_if_t<std::is_floating_point<T2>::value, T>
+  fetch_add(T Operand, memory_order Order = memory_order::relaxed) {
+    return __spirv_AtomicFAdd(
+        Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
+  }
+
+  template <typename T2 = T>
+  detail::enable_if_t<std::is_integral<T2>::value, T>
+  fetch_sub(T Operand, memory_order Order = memory_order::relaxed) {
     __SYCL_STATIC_ASSERT_NOT_FLOAT(T);
     return __spirv_AtomicISub(
         Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
   }
 
+  template <typename T2 = T>
+  detail::enable_if_t<std::is_floating_point<T2>::value, T>
+  fetch_sub(T Operand, memory_order Order = memory_order::relaxed) {
+    // Negate the float add instruction.
+    return -__spirv_AtomicFAdd(
+        Ptr, SpirvScope, detail::getSPIRVMemorySemanticsMask(Order), Operand);
```

Comment:
Should the negation here apply to `Operand` instead of the result?

Reply:
It **definitely** should, thanks for noting this :) Fixed.

That's what leaving a PR unattended for a month will do to me!